modules/core/core-event/src/main/java/com/enonic/xp/core/impl/event/cluster/ClusterEventReceiver.java

```diff
@@ -0,0 +1,54 @@
+package com.enonic.xp.core.impl.event.cluster;
+
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Deactivate;
+import org.osgi.service.component.annotations.Reference;
+
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.core.ITopic;
+import com.hazelcast.core.Message;
+import com.hazelcast.core.MessageListener;
+
+import com.enonic.xp.event.Event;
+import com.enonic.xp.event.EventPublisher;
+
+@Component(immediate = true)
+public final class ClusterEventReceiver
+    implements MessageListener<Event>
+{
+    private final EventPublisher eventPublisher;
+
+    private final ITopic<Event> topic;
+
+    private String registrationId;
+
+    @Activate
+    public ClusterEventReceiver( @Reference final HazelcastInstance hazelcastInstance, @Reference final EventPublisher eventPublisher )
+    {
+        this.topic = hazelcastInstance.getTopic( ClusterEventSender.ACTION );
+        this.eventPublisher = eventPublisher;
+    }
+
+    @Activate
```

Comment:
@rymsha why do you use two times `@Activate` annotation?

Reply:
One for "constructor injection", another for real activation logic.